compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirExpressionsResolveTransformer.kt

```diff
@@ -188,44 +194,14 @@ class FirExpressionsResolveTransformer(transformer: FirBodyResolveTransformer) :
             }
 
         dataFlowAnalyzer.exitFunctionCall(completeInference, callCompleted)
-        if (callCompleted && completeInference.isArrayOfCall) {
-            return completeInference.toArrayOfCall().compose()
-        }
-        return completeInference.compose()
-    }
-
-    private val FirFunctionCall.isArrayOfCall: Boolean
-        get() {
-            val function: FirCallableDeclaration<*> = getOriginalFunction() ?: return false
-            return function is FirSimpleFunction &&
-                    // TODO: extend it to other variants, like emptyArray, intArrayOf, floatArrayOf, etc.
-                    function.name.asString() == "arrayOf" &&
-                    function.returnTypeRef.isArrayType &&
-                    function.valueParameters.size == 1 && function.valueParameters[0].isVararg &&
-                    arguments.size == 1 &&
-                    function.receiverTypeRef == null
-        }
-
-    private fun FirFunctionCall.toArrayOfCall(): FirArrayOfCall {
-        assert(isArrayOfCall) {
-            "Unexpected transformation from $this to FirArrayOfCall"
-        }
-        val functionCall = this
-        val typeRef = this.typeRef
-        return buildArrayOfCall {
-            source = functionCall.source
-            annotations += functionCall.annotations
-            // Note that the signature is: arrayOf(vararg element). Hence, unwrapping the original argument list here.
-            argumentList = buildArgumentList {
-                if (functionCall.arguments.isNotEmpty()) {
-                    (functionCall.argument as FirVarargArgumentsExpression).arguments.forEach {
-                        arguments += it
-                    }
+        if (callCompleted) {
+            if (enableArrayOfCallTransformation) {
```

Comment:
Note that I intentionally split these conditions, since I'm working on another transformer (constant evaluator that will transform resolved numeric calls with const expressions into const expression) that will be added around here.